besom-cfg/lib/src/main/scala/Configured.scala

```diff
@@ -26,75 +31,84 @@ import besom.cfg.internal.*
-          case '[t] =>
-            Expr.summon[ConfiguredType[t]].getOrElse {
-              report.error(
-                s"Cannot find ConfiguredType for type ${tpe.memberType(sym).dealias.show}"
-              )
-              throw new Exception("Cannot find ConfiguredType")
-            }
-          case _ =>
-            report.error("Unsupported type")
-            throw new Exception("Unsupported type")
-
-      '{ Field(${ name }, ${ ftpe }.toFieldType) }
-    }
-
-    val fromEnvExpr = Expr.summon[FromEnv[A]].getOrElse {
-      report.error(s"Cannot find FromEnv for type ${tpe.show}")
-      throw new Exception("Cannot find FromEnv")
-    }
-
-    val schemaExpr = '{ Schema(${ Expr.ofList(fields) }.toList, ${ Expr(Version) }) }
-
-    '{
-      new Configured[A] {
-        def schema = $schemaExpr
-        def newInstanceFromEnv(env: Map[String, String] = sys.env): A =
-          $fromEnvExpr.decode(env, "").getOrElse {
-            throw new Exception("Failed to decode")
-          }
+
+  trait FromEnv[A] extends Configured[A, Map[String, String]]:
+    def schema: Schema
+    def newInstance(input: Map[String, String]): A
+
+  object FromEnv:
+    val MediumIdentifier = "env"
+
+    inline def derived[A <: Product]: Configured.FromEnv[A] = ${ derivedImpl[A] }
+
+    def derivedImpl[A <: Product: Type](using ctx: Quotes): Expr[Configured.FromEnv[A]] =
+      import ctx.reflect.*
+
+      val tpe = TypeRepr.of[A]
+      val fields = tpe.typeSymbol.caseFields.map { case sym =>
+        val name = Expr(sym.name)
+        val ftpe: Expr[ConfiguredType[_]] =
+          tpe.memberType(sym).dealias.asType match
+            case '[t] =>
+              Expr.summon[ConfiguredType[t]].getOrElse {
+                report.errorAndAbort(
+                  s"Cannot find ConfiguredType for type ${tpe.memberType(sym).dealias.show}"
+                )
+              }
+
+            case _ => report.errorAndAbort("Unsupported type")
+
+        '{ Field(${ name }, ${ ftpe }.toFieldType) }
+      }
+
+      val fromEnvExpr = Expr.summon[from.env.ReadFromEnvVars[A]].getOrElse {
+        report.error(s"Cannot find FromEnv for type ${tpe.show}")
+        throw new Exception("Cannot find FromEnv")
+      }
+
+      val schemaExpr = '{ Schema(${ Expr.ofList(fields) }.toList, ${ Expr(Version) }, ${ Expr(FromEnv.MediumIdentifier) }) }
+
+      '{
+        new Configured.FromEnv[A] {
+          def schema = $schemaExpr
+          def newInstance(input: Map[String, String]): A =
+            $fromEnvExpr.decode(input, from.env.EnvPath.Root) match
+              case Validated.Valid(a)        => a
+              case Validated.Invalid(errors) => throw ConfigurationError(errors.toVector)
+
+        }
       }
-    }
+    end derivedImpl
+  end FromEnv
 end Configured
 
-def resolveConfiguration[A](using c: Configured[A]): A =
-  c.newInstanceFromEnv()
+def resolveConfiguration[A, INPUT: Default](using c: Configured[A, INPUT]): A =
```

Comment:
Is this supposed to be proper API for users?
Should they then call something like `resolveConfiguration[Foo, Map[String, String]]`?
Are we sure this is going to work well when we mix type inference and implicit resolution with overloading?

Reply:
this is a good question and I think this was WIP but since we're killing macros-based approach we can ignore this and figure that out later, no?

Reply:
ok

Reply:
fixed